flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java

```diff
@@ -109,7 +119,11 @@ public MiniDispatcher(
 
 		if (executionMode == ClusterEntrypoint.ExecutionMode.NORMAL) {
 			// terminate the MiniDispatcher once we served the first JobResult successfully
-			jobResultFuture.whenComplete((JobResult ignored, Throwable throwable) -> shutDown());
+			jobResultFuture.whenComplete((JobResult result, Throwable throwable) -> {
+				ApplicationStatus status = result.getSerializedThrowable().isPresent() ?
+						ApplicationStatus.FAILED : ApplicationStatus.SUCCEEDED;
+				jobTerminationFuture.complete(status);
```

Comment:
I think the functional way would be:

```
				jobTerminationFuture.complete(result.getSerializedThrowable()
					.map(serializedThrowable -> ApplicationStatus.FAILED)
					.orElse(ApplicationStatus.SUCCEEDED));
```

Reply:
True, I had it like that initially, but found the above version more readable in the end, because we don't really use the serializedThrowable (making the map() a bit strange).